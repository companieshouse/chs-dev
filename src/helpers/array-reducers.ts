export const collect = <U, T>(fn: (value: T) => U[]): (result: U[], value: T) => U[] => {
    return (values: U[], value: T): U[] => {
        values.push(...fn(value));
        return values;
    };
};

export const deduplicate = (uniqueValues: string[], value: string): string[] => {
    if (!uniqueValues.includes(value)) {
        uniqueValues.push(value);
    }
    return uniqueValues;
};
