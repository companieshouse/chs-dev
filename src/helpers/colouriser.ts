import { cyan, gray, green, red, yellow } from "ansis";

type ColourMap = [RegExp, (str: string) => string][];

const statusColourMap: ColourMap = [
    [/Exited\s\([1-9]|\d{2,}\)/, red],
    [/Exited\s\(0\)/, yellow],
    [/Up[^()]+$/, green],
    [/Up.+\(health:\sstarting./, cyan],
    [/Up.+\(healthy./, green.bold],
    [/Up.+\(unhealthy/, red.bold],
    [/Not\srunning/, gray]
];

const runStatusColourMap: ColourMap = [
    [/Started/, green],
    [/Healthy/, green.bold],
    [/Stopped/, yellow]
];

const stopStatusColourMap: ColourMap = [
    [/Stopped/, yellow],
    [/Removed/, red]
];

const colourise = (colourMap: ColourMap) => (str: string) => {
    const colour = colourMap
        .find(([regex, _]) => regex.test(str));

    if (colour) {
        return colour[1](str);
    } else {
        return str;
    }
};

type SimpleColour = "cyan" | "red" | "bold-red" | "green" | "bold-green" | "yellow" | "gray" | "grey"

export const simpleColouriser = (str: string, colour: SimpleColour) => {
    let colouriseFn: (str: string) => string;

    switch (colour) {
    case "cyan":
        colouriseFn = cyan;
        break;
    case "red":
        colouriseFn = red;
        break;
    case "bold-red":
        colouriseFn = red.bold;
        break;
    case "green":
        colouriseFn = green;
        break;
    case "bold-green":
        colouriseFn = green.bold;
        break;
    case "yellow":
        colouriseFn = yellow;
        break;
    case "gray":
    case "grey":
        colouriseFn = gray;
        break;
    default:
        colouriseFn = (strToColourise: string) => strToColourise;
        break;
    }

    return colouriseFn(str);
};

export const statusColouriser = colourise(statusColourMap);
export const runStatusColouriser = colourise(runStatusColourMap);
export const stopStatusColouriser = colourise(stopStatusColourMap);
