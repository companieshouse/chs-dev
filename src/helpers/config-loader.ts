import { existsSync, readFileSync } from "fs";
import { basename, join } from "path";
import yaml from "yaml";
import Config from "../model/Config.js";
import Constants from "../model/Constants.js";
import { getDockerSettingsFilePath } from "./docker-settings-config.js";

const fileVarRegExp = /^file:\/\/(.+)$/;

/**
 * Loads the configuration from the project root. When project does not contain
 * configuration then returns an empty configuration object.
 * @returns Project Config
 */
export const load: () => Config = () => {
    const projectPath = process.env.CHS_DEV_PROJECT || process.cwd();
    const confFile = join(projectPath, "chs-dev/config.yaml");
    const dockerSettingsPath:string = getDockerSettingsFilePath();

    let config: Partial<Config> = {};

    if (!existsSync(confFile)) {
        config = {
            env: {}
        };

    } else {
        const rawConfiguration = readFileSync(join(confFile)).toString("utf8");

        const projectConfiguration = yaml.parse(rawConfiguration);

        config = {
            env: parseEnv(projectConfiguration.env) || {},
            performEcrLoginHoursThreshold: projectConfiguration.authed_repositories && projectConfiguration.ecr_login_threshold_hours ? projectConfiguration.ecr_login_threshold_hours : undefined,
            versionSpecification: projectConfiguration.version
        };
    }

    if (!config.performEcrLoginHoursThreshold) {
        config = {
            ...config,
            performEcrLoginHoursThreshold: Constants.DEFAULT_PERFORM_ECR_LOGIN_HOURS_THRESHOLD
        };
    }

    return {
        ...config,
        projectPath,
        projectName: basename(projectPath),
        dockerSettingsPath
    } as Config;
};

const parseEnv = (env: Record<string, string> | undefined) => {
    if (!env) {
        return undefined;
    }

    return Object.keys(env)
        .map(key => {
            const value = env[key];

            const fileMatches = fileVarRegExp.exec(value);

            if (fileMatches) {
                const file = fileMatches[1];

                if (file && existsSync(file)) {
                    return [key, readFileSync(file).toString("utf8")];
                }
            }

            return [key, value];
        })
        .reduce((previous, next) => {
            const [key, value] = next;

            previous[key] = value;

            return previous;
        }, {} as Record<string, string>);
};

export default load;
