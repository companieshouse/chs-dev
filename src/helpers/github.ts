import { execSync } from "child_process";
import CONSTANTS from "../model/Constants.js";
import { displayLink } from "./link.js";
import { spawn } from "./spawn-promise.js";

/**
 * @returns the personal access token for the GitHub API or undefined if
 * could not be found.
 */
export const getPersonalAccessToken = async () => {
    return process.env.GITHUB_PAT || await attemptToGetFromGhCli();
};

/**
 * Creates a Pull Request for the given repository, branch and title.
 * @param repository Name of the repository
 * @param branch Name of the branch
 * @param title Title of the Pull Request
 * @param destinationBranch Name of the destination branch
 * @param organisation name of the GitHub organisation
 * @returns Promise<void> indicating the completion of the operation
 */
export const createPullRequest = async (
    repository: string,
    branch: string,
    title: string,
    destinationBranch: string | undefined = undefined,
    organisation: string | undefined = undefined
) => {
    const gitHibOrganisationName = organisation || CONSTANTS.COMPANIES_HOUSE_GITHUB_ORGANISATION_NAME;
    const githubApiToken = await getPersonalAccessToken();

    if (typeof githubApiToken !== "undefined") {
        try {
            const result = await fetch(`https://api.github.com/repos/${gitHibOrganisationName}/${repository}/pulls`, {
                method: "POST",
                headers: {
                    Authorization: `Bearer ${githubApiToken}`,
                    "Content-Type": "application/json"
                },
                body: JSON.stringify({
                    title,
                    head: branch,
                    base: destinationBranch || "main"
                })
            });

            if (result.status >= 200 && result.status < 300) {
                const apiResponse = await result.json();

                console.log(`Pull request created for ${repository} with title: ${title} id: ${apiResponse.number}`);

                return;
            }
        } catch (error) {
            console.error("Failed to create pull request, you'll have to create it using your web browser", error);
        }
    }

    openPullRequestInBrowser(gitHibOrganisationName, repository, destinationBranch, branch, title);
};

/**
 * Fetches the GitHub repository description using the gh CLI.
 * @param repository
 * @param muteError
 * @returns returned as a string, the description of the repository
 */
export const getRepositoryDescription = (repositoryName: string, muteError = false): string => {
    try {
        const stdout = execSync(`
            gh repo view --json 'description' companieshouse/${repositoryName} 2>/dev/null`).toString("utf8");

        const parsed = JSON.parse(stdout);
        return parsed.description ?? "No description available.";
    } catch (error) {
        if (!muteError) {
            console.error(`Failed to fetch description for ${repositoryName}:`, error);
            return "Error fetching description.";
        }
        return "No description available.";
    }
};

/**
  * Fetches the repository code_owner using the gh CLI.
 * @param repository
 * @param muteError
 * @returns team-code-owner value
 */
export const getRespositoryOwner = (repositoryName: string, muteError = false): string => {
    try {
        const stdout = execSync(`gh api /repos/companieshouse/${repositoryName} --jq '.custom_properties["team-code-owner"]' 2>/dev/null`).toString("utf8");

        return stdout || "Unknown repository owner.";
    } catch (error) {
        if (!muteError) {
            console.error(`Failed to fetch team code owner for ${repositoryName}:`, error);
            return "Error fetching team code owner.";
        }
        return "Unknown repository owner.";
    }
};

const attemptToGetFromGhCli = async () => {
    try {
        await spawn("command", ["-v", "gh"], {
            logHandler: {
                handle: () => { }
            }
        });

        const tokenReader = new TokenReaderLogHandler();
        tokenReader.reset();

        await spawn("gh", ["auth", "token"], {
            logHandler: tokenReader
        });

        return tokenReader.tokenValue;
    } catch (error) {
        return undefined;
    }
};

class TokenReaderLogHandler {
    private token: string;

    constructor () {
        this.token = "";
    }

    handle (logEntry: string) {
        this.token += logEntry;
    }

    reset () {
        this.token = "";
    }

    get tokenValue () {
        return this.token;
    }
}

const openPullRequestInBrowser = (
    gitHibOrganisationName: string,
    repository: string,
    destinationBranch: string | undefined,
    branch: string,
    title: string
) => {
    displayLink(
        `https://github.com/${gitHibOrganisationName}/${repository}/compare/${destinationBranch || "main"}...${branch}?expand=1&title=${title}`,
        "Create Pull Request (In browser)"
    );
};
