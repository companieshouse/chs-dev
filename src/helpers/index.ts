/**
 * Checks if a given value is an object and not an array.
 *
 * @param {unknown} T - The value to check.
 * @returns {boolean} - Returns `true` if the value is an object and not an array, otherwise `false`.
 */
export const isObjectFormat = (T: unknown): boolean => {
    return typeof T === "object" && !Array.isArray(T) && T !== null;
};

/**
 * Determines if an object is empty.
 *
 * @param {Record<string, any>} obj - The object to check.
 * @returns {boolean} - Returns `true` if the object has no own enumerable properties, otherwise `false`.
 */
export const isEmptyObject = (obj: Record<string, any>): boolean => {
    return Object.keys(obj).length === 0;
};

/**
 * Checks if there is at least one common element between two arrays.
 *
 * @param {unknown[]} array1 - The first array to compare.
 * @param {unknown[]} array2 - The second array to compare.
 * @returns {boolean} - Returns `true` if any element in `array1` exists in `array2`, otherwise `false`.
 */
export const matchExistInArrays = (array1: unknown[], array2: unknown[]): boolean => {
    return array1.some(item => array2.includes(item));
};

/**
 * Finds and returns elements from the parentArray that do not exist in the childArray.
 *
 * @param {unknown[]} parentArray - The source array to check.
 * @param {unknown[]} childArray - The reference array to compare against.
 * @returns {any[]} - An array containing elements from `parentArray` that are not present in `childArray`.
 */
export const findUniqueItemsInParentArray = (parentArray: unknown[], childArray: unknown[]): any[] => {
    const childSet = new Set(childArray);
    return parentArray.filter(item => !childSet.has(item));
};
