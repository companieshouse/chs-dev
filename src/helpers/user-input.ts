import * as inquirer from "@inquirer/prompts";

type ConfirmOptions = {
    defaultValue?: boolean
}

type Choice = {
    value: string;
    name?: string;
    description?: string;
    short?: string;
    disabled?: boolean | string
}

type InputOptions = {
    options?: string[] | Choice[],
    defaultValue?: string
}

type EditorOptions = {
    initialValue?: string,
    waitForUserInput?: boolean
}

const defaultValue = (opts: any | undefined) => ({
    ...(typeof opts !== "undefined" && typeof opts.defaultValue !== "undefined" ? { default: opts.defaultValue } : {})
});

export const confirm = (
    question: string,
    options?: ConfirmOptions
) => {
    return inquirer.confirm({
        message: question,
        ...defaultValue(options)
    });
};

export const input: (q: string, o?: InputOptions) => Promise<string> = (
    question: string,
    options?: InputOptions
) => {

    if (typeof options?.options === "undefined") {
        return inquirer.input({
            message: question,
            ...defaultValue(options)
        });
    } else {
        return inquirer.select({
            message: question,
            choices: options.options.some(option => typeof option === "string")
                ? options.options.map(optionValue => ({ value: optionValue }))
                : options.options as Choice[],
            ...defaultValue(options)
        });
    }
};

export const editor: (
    q: string,
    o?: EditorOptions
) => Promise<string> = (
    question: string,
    options?: EditorOptions
): Promise<string> => {
    return inquirer.editor({
        message: question,
        ...defaultValue({ defaultValue: options?.initialValue }),
        ...(typeof options?.waitForUserInput !== "undefined"
            ? {
                waitForUserInput: options.waitForUserInput
            }
            : {})
    });
};
