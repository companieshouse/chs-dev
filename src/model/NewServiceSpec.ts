type NewServiceSpec = {
    name: string,

    description: string

    type: {
        name: "MICROSERVICE",
        language: "JAVA" | "NODE"
    },

    configuration: {
        java?: {
            spring_initializr_url: string
        },
        ["docker-chs-development"]?: {
            module: string
        },
        deploy?: {
            stack_name: string
        }
    },

    submission_details: {
        by: string,
        github_username: string
    },

    ownership: {
        team: string,

        service: string
    },

    sensitive: boolean,

    sensitivity_justification?: string
}

export default NewServiceSpec;
