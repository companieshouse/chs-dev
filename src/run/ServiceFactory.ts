import serviceSpecPrompts from "./service-factory/prompts/index.js";
import buildServiceSpecFromUserInput from "./service-factory/build-service-spec.js";
import ServiceSpecRepository from "./service-factory/ServiceSpecRepository.js";
import { confirm } from "../helpers/user-input.js";
import NewServiceSpec from "../model/NewServiceSpec.js";
import yaml from "yaml";
import CONSTANTS from "../model/Constants.js";
import { createPullRequest } from "../helpers/github.js";
import { readFileSync } from "fs";

/**
 * ServiceFactory is responsible for creating new service specifications.
 */
export default class ServiceFactory {

    // eslint-disable-next-line no-useless-constructor
    constructor (private readonly serviceSpecRepository: ServiceSpecRepository) { }

    /**
     * For the supplied Service Name will create a new service specification. It
     * will initialise the repository and commit the new service specification
     * to its own branch before raising a pull request. However, it will not
     * perform the commit or pull request if the dryRun flag is set to true.
     * @param serviceName Name of the service to create
     * @param dryRun whether to perform the operation or just simulate it
     * @param file containing the service specification
     * @returns Promise<void> indicating the completion of the operation
     */
    async createNewService (serviceName: string, dryRun: boolean = false, file: string | undefined = undefined): Promise<void> {
        await this.serviceSpecRepository.load();

        try {
            const branchName = this.createBranchNameForService(serviceName);

            await this.serviceSpecRepository.initialise(branchName);

            const initialServiceSpec = await this.serviceSpecRepository.get(serviceName);

            const overrideServiceSpec: Partial<NewServiceSpec> | undefined = this.readInputFileServiceSpec(file);

            if (typeof initialServiceSpec !== "undefined") {
                if (!await confirm(`Are you sure you want to modify the existing service specification for service: ${serviceName}?`)) {
                    throw new Error(`The service: ${serviceName} already exists and was not modified.`);
                }
            }

            const builtServiceSpec = await this.buildServiceSpec(overrideServiceSpec, initialServiceSpec, serviceName);

            await this.serviceSpecRepository.put(serviceName, builtServiceSpec as NewServiceSpec);

            await this.submitNewServiceChanges(serviceName, builtServiceSpec, branchName, dryRun);
        } finally {
            this.serviceSpecRepository.close();
        }
    }

    private async submitNewServiceChanges (
        serviceName: string,
        builtServiceSpec: Partial<NewServiceSpec>,
        branchName: string,
        dryRun: boolean
    ) {
        if (dryRun) {
            this.displayDryRunServiceSpec(builtServiceSpec);

            return;
        }

        await this.serviceSpecRepository.commitAndPush(
            `Add new service: ${serviceName}`,
            `* Create new service specification for ${serviceName}`,
            "```json",
            JSON.stringify(builtServiceSpec, null, 2),
            "```"
        );

        await createPullRequest(
            CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_NAME,
            branchName,
            `Create new service specification for ${serviceName}`,
            CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_BRANCH
        );
    }

    private displayDryRunServiceSpec (builtServiceSpec: Partial<NewServiceSpec>) {
        console.log(
            `(dry-run) Would have committed and pushed ${CONSTANTS.NEW_SERVICE_SPECS_PATH}/${builtServiceSpec.name}.yaml with the following content:`
        );

        this.printCreatedNewServiceSpec(builtServiceSpec as NewServiceSpec);

        console.log(
            `(dry-run) would have opened a pull request to ${CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_NAME} into branch ${CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_BRANCH}`
        );
    }

    private async buildServiceSpec (
        overrideServiceSpec: Partial<NewServiceSpec> | undefined,
        initialServiceSpec: NewServiceSpec | undefined,
        serviceName: string
    ) {
        const serviceSpecBuilder = buildServiceSpecFromUserInput(serviceSpecPrompts);

        const builderInputAsInitialValue = {
            ...(initialServiceSpec || {
                name: serviceName
            })
        };

        const builderInput = typeof overrideServiceSpec === "undefined"
            ? builderInputAsInitialValue
            : {
                ...builderInputAsInitialValue,
                ...overrideServiceSpec
            };

        return await serviceSpecBuilder(builderInput);
    }

    private readInputFileServiceSpec (file: string | undefined): NewServiceSpec | undefined {
        let overrideServiceSpec: NewServiceSpec | undefined;

        if (typeof file !== "undefined") {
            overrideServiceSpec = yaml.parse(readFileSync(file).toString("utf-8"));

            if (typeof overrideServiceSpec !== "object") {
                throw new Error("Unable to parse yaml file");
            }

            const invalidOverrideKeys = this.listInvalidOverrideKeys(overrideServiceSpec);

            if (invalidOverrideKeys.length > 0) {
                throw new Error("Invalid override file, contains invalid keys: " + invalidOverrideKeys.join(", "));
            }
        }

        return overrideServiceSpec;
    }

    private createBranchNameForService (serviceName: string) {
        const date = new Date(Date.now());

        const dateLabel = date.toLocaleDateString(
            "en-CA", {
                year: "numeric",
                month: "2-digit",
                day: "2-digit"
            }
        );

        return `feature/add-${serviceName.replace(/[^a-zA-Z0-9\-.]/g, "-").toLowerCase()}-${dateLabel}`;
    }

    private printCreatedNewServiceSpec (newServiceSpec: NewServiceSpec) {
        console.log(yaml.stringify(newServiceSpec));
    }

    private listInvalidOverrideKeys (overrideServiceSpec: Partial<NewServiceSpec>) {
        // Since name is provided by the user, it is a valid key
        const validPaths = [...serviceSpecPrompts.map(prompt => prompt.selector), ".name"];

        // Flatten the object to a list of paths
        const flattenObject = (initialPath: string, object: object): [string, any][] => {
            return Object.entries(object)
                .flatMap(([key, value]) => {
                    const path = initialPath === "." ? key : `${initialPath}.${key}`;

                    return typeof value === "object" ? flattenObject(path, value) : [[path, value]];
                });
        };

        return flattenObject(".", overrideServiceSpec)
            .filter(([path, value]) => typeof value !== "undefined" && !validPaths.includes(`.${path}`))
            .map(([path, _]) => path);
    }

}
