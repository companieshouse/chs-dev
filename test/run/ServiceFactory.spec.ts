import { expect, jest, test } from "@jest/globals";
import buildServiceSpecFactoryMock from "../../src/run/service-factory/build-service-spec";
import ServiceFactory from "../../src/run/ServiceFactory";
import { confirm as confirmMock } from "../../src/helpers/user-input";
import NewServiceSpec from "../../src/model/NewServiceSpec";
import yaml from "yaml";
import CONSTANTS from "../../src/model/Constants";
import { createPullRequest as createPullRequestMock } from "../../src/helpers/github";
import fsMock from "fs";

jest.mock("../../src/run/service-factory/build-service-spec");
jest.mock("../../src/helpers/user-input");
jest.mock("../../src/helpers/github");
jest.mock("fs");

const serviceSpecRepositoryMock = {
    load: jest.fn(),
    initialise: jest.fn(),
    get: jest.fn(),
    put: jest.fn(),
    commitAndPush: jest.fn(),
    close: jest.fn()
};

const newServiceSpec: NewServiceSpec = {
    name: "my-new-service",
    type: {
        name: "MICROSERVICE",
        language: "JAVA"
    },
    description: "My new service",
    configuration: {
        java: {
            spring_initializr_url: "https://start.spring.io#!p=134"
        },
        "docker-chs-development": {
            module: "module"
        }
    },
    submission_details: {
        by: "Test User",
        github_username: "ch-test-user"
    },
    ownership: {
        team: "Element",
        service: "Common Component"
    },
    sensitive: true,
    sensitivity_justification: "It's sensitive alright!"
};

const testDateTime = new Date(2025, 0, 8, 0, 0, 0);

describe("ServiceFactory", () => {

    let serviceFactory: ServiceFactory;

    const buildServiceSpecMock = jest.fn();

    beforeEach(() => {
        jest.resetAllMocks();
        serviceFactory = new ServiceFactory(serviceSpecRepositoryMock as never);

        // @ts-expect-error  This is mocked and so the function should be there
        buildServiceSpecFactoryMock.mockReturnValue(buildServiceSpecMock);

        // @ts-expect-error
        Date.now = () => testDateTime;
    });

    it("loads the repository", async () => {
        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.load).toHaveBeenCalled();
    });

    it("initialises the repository", async () => {
        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.initialise).toHaveBeenCalledWith("feature/add-my-new-service-2025-01-08");
    });

    test.each([
        ["My_New_Service", "my-new-service"],
        ["My_New_Service234", "my-new-service234"],
        ["My&New&Service", "my-new-service"],
        ["My.New.&Service", "my.new.-service"]
    ])("initialises replaces %s with %s in branch name", async (serviceName, expectedServiceNameInBranch) => {

        await serviceFactory.createNewService(serviceName);

        expect(serviceSpecRepositoryMock.initialise).toHaveBeenCalledWith(`feature/add-${expectedServiceNameInBranch}-2025-01-08`);
    });

    it("gets the initial state of the service", async () => {
        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.get).toHaveBeenCalledWith("my-new-service");
    });

    it("closes the repository when complete", async () => {
        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.close).toHaveBeenCalled();
    });

    it("constructs the new service spec", async () => {
        serviceSpecRepositoryMock.get.mockReturnValue(undefined);

        await serviceFactory.createNewService("my-new-service");

        expect(buildServiceSpecMock).toHaveBeenCalledWith({
            name: "my-new-service"
        });
    });

    it("asks user to confirm amendment when service spec already exists", async () => {
        (confirmMock as jest.MockedFunction<typeof confirmMock>).mockResolvedValue(true);

        serviceSpecRepositoryMock.get.mockReturnValue({
            name: "service"
        });

        await serviceFactory.createNewService("service");

        expect(confirmMock).toHaveBeenCalledWith(
            "Are you sure you want to modify the existing service specification for service: service?"
        );
    });

    it("passes service to service spec builder when user accepts", async () => {
        (confirmMock as jest.MockedFunction<typeof confirmMock>).mockResolvedValue(true);

        serviceSpecRepositoryMock.get.mockReturnValue({
            name: "service",
            description: "service description"
        });

        await serviceFactory.createNewService("service");

        expect(buildServiceSpecMock).toHaveBeenCalledWith({
            name: "service",
            description: "service description"
        });
    });

    it("rejects if the user does not want to overwrite the existing service spec", async () => {
        (confirmMock as jest.MockedFunction<typeof confirmMock>).mockResolvedValue(false);

        serviceSpecRepositoryMock.get.mockReturnValue({
            name: "service",
            description: "service description"
        });

        await expect(serviceFactory.createNewService("service")).rejects.toEqual(
            new Error("The service: service already exists and was not modified.")
        );
    });

    it("closes the repository when there is an error", async () => {
        (confirmMock as jest.MockedFunction<typeof confirmMock>).mockResolvedValue(false);

        serviceSpecRepositoryMock.get.mockReturnValue({
            name: "service",
            description: "service description"
        });

        await expect(serviceFactory.createNewService("service")).rejects.toBeDefined();

        expect(serviceSpecRepositoryMock.close).toHaveBeenCalled();
    });

    it("puts the new service spec into the repository", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.put).toHaveBeenCalledWith("my-new-service", newServiceSpec);
    });

    it("commits and pushes the new service spec to the repository", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service");

        expect(serviceSpecRepositoryMock.commitAndPush).toHaveBeenCalledWith(
            "Add new service: my-new-service",
            "* Create new service specification for my-new-service",
            "```json",
            JSON.stringify(newServiceSpec, null, 2),
            "```"
        );
    });

    it("does not commit and push when is dry run", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service", true);

        expect(serviceSpecRepositoryMock.commitAndPush).not.toHaveBeenCalled();
    });

    it("closes the repository when complete dry run", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service", true);

        expect(serviceSpecRepositoryMock.close).toHaveBeenCalled();
    });

    it("prints out the created service spec on dry run", async () => {
        const consoleLogSpy = jest.spyOn(console, "log");

        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service", true);

        expect(consoleLogSpy).toBeCalledWith(
            `(dry-run) Would have committed and pushed ${CONSTANTS.NEW_SERVICE_SPECS_PATH}/my-new-service.yaml with the following content:`
        );
        expect(consoleLogSpy).toHaveBeenCalledWith(yaml.stringify(newServiceSpec));
    });

    it("creates a pull request when not a dry run", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service");

        expect(createPullRequestMock).toHaveBeenCalledWith(
            CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_NAME,
            "feature/add-my-new-service-2025-01-08",
            "Create new service specification for my-new-service",
            CONSTANTS.NEW_SERVICE_SPECS_REPOSITORY_BRANCH
        );
    });

    it("does not attempt to create a pull request when is dry run", async () => {
        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service", true);

        expect(createPullRequestMock).not.toHaveBeenCalled();
    });

    it("reads supplied file and creates a new service spec", async () => {
        const newServiceSpecOverrides = {
            name: "my-new-service",
            type: {
                name: "MICROSERVICE"
            }
        };

        (fsMock.readFileSync as jest.Mock).mockReturnValue(
            Buffer.from(yaml.stringify(newServiceSpecOverrides), "utf-8")
        );

        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await serviceFactory.createNewService("my-new-service", true, "path/to/file.yaml");

        expect(fsMock.readFileSync).toHaveBeenCalledWith("path/to/file.yaml");

        expect(buildServiceSpecMock).toHaveBeenCalledWith({
            ...newServiceSpecOverrides
        });
    });

    it("throws error when yaml file is not valid", async () => {
        (fsMock.readFileSync as jest.Mock).mockReturnValue(
            Buffer.from("invalid yaml", "utf-8")
        );

        await expect(serviceFactory.createNewService("my-new-service", true, "path/to/file.yaml")).rejects.toEqual(
            new Error("Unable to parse yaml file")
        );
    });

    it("raises error when file contains invalid keys", async () => {
        const newServiceSpecOverrides = {
            name: "my-new-service",
            type: {
                name: "MICROSERVICE"
            },
            invalidKey: "invalid",
            anotherInvalidKey: true
        };

        (fsMock.readFileSync as jest.Mock).mockReturnValue(
            Buffer.from(yaml.stringify(newServiceSpecOverrides), "utf-8")
        );

        buildServiceSpecMock.mockResolvedValue(newServiceSpec as never);

        await expect(serviceFactory.createNewService("my-new-service", true, "path/to/file.yaml")).rejects.toMatchSnapshot();
    });
});
