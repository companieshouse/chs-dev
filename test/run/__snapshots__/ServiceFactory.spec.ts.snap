// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceFactory raises error when file contains invalid keys 1`] = `[Error: Invalid override file, contains invalid keys: invalidKey, anotherInvalidKey]`;
